runtime/src/xcm_tests/mod.rs

```diff
@@ -0,0 +1,203 @@
+
+decl_test_network! {
+	pub struct TestNet {
+		relay_chain = Relay,
+		parachains = vec![
+			(1, ParaX),
+			(2, SoraParachain),
+		],
+	}
+}
+
+decl_test_relay_chain! {
+	pub struct Relay {
+		Runtime = relay::Runtime,
+		XcmConfig = relay::XcmConfig,
+		new_ext = relay_ext(),
+	}
+}
+
+decl_test_parachain! {
+	pub struct ParaX {
+		Runtime = para_x::Runtime,
+		XcmpMessageHandler = para_x::XcmpQueue,
+		DmpMessageHandler = para_x::DmpQueue,
+		new_ext = para_ext(1),
+	}
+}
+
+decl_test_parachain! {
+	pub struct SoraParachain {
+		Runtime = crate::Runtime,
+		XcmpMessageHandler = crate::XcmpQueue,
+		DmpMessageHandler = crate::DmpQueue,
+		new_ext = para_ext(2),
+	}
+}
+
+// Configure Mock Relay Chain
+
+pub fn relay_ext() -> sp_io::TestExternalities {
+	use relay::{Runtime, System};
+
+	let mut t = frame_system::GenesisConfig::default().build_storage::<Runtime>().unwrap();
+
+	pallet_balances::GenesisConfig::<Runtime> { balances: vec![(ALICE, 1_000)] }
+		.assimilate_storage(&mut t)
+		.unwrap();
+
+	let mut ext = sp_io::TestExternalities::new(t);
+	ext.execute_with(|| System::set_block_number(1));
+	ext
+}
+
+// Configure Mock dummy Parachain:
+
+pub type Balance = u128;
+pub type Amount = i128;
+
+pub struct AllTokensAreCreatedEqualToWeight(MultiLocation);
+impl WeightTrader for AllTokensAreCreatedEqualToWeight {
+	fn new() -> Self {
+		Self(MultiLocation::parent())
+	}
+
+	fn buy_weight(&mut self, weight: Weight, payment: Assets) -> Result<Assets, XcmError> {
+		let asset_id = payment.fungible.iter().next().expect("Payment must be something; qed").0;
+		let required = MultiAsset { id: asset_id.clone(), fun: Fungible(weight as u128) };
+
+		if let MultiAsset { fun: _, id: Concrete(ref id) } = &required {
+			self.0 = id.clone();
+		}
+
+		let unused = payment.checked_sub(required).map_err(|_| XcmError::TooExpensive)?;
+		Ok(unused)
+	}
+
+	fn refund_weight(&mut self, weight: Weight) -> Option<MultiAsset> {
+		if weight.is_zero() {
+			None
+		} else {
+			Some((self.0.clone(), weight as u128).into())
+		}
+	}
+}
+
+pub fn para_ext(para_id: u32) -> TestExternalities {
+	use para_x::{Runtime, System};
+
+	let mut t = frame_system::GenesisConfig::default().build_storage::<Runtime>().unwrap();
+
+	let parachain_info_config = parachain_info::GenesisConfig { parachain_id: para_id.into() };
+	<parachain_info::GenesisConfig as frame_support::traits::GenesisBuild<Runtime, _>>::assimilate_storage(&parachain_info_config, &mut t)
+		.unwrap();
+
+	orml_tokens::GenesisConfig::<Runtime> {
+		balances: vec![
+			(ALICE, CurrencyId::R, 1_000_000_000_000_000_000),
+			(ALICE, CurrencyId::X, 1_000_000_000_000_000_000),
+		],
+	}
+	.assimilate_storage(&mut t)
+	.unwrap();
+
+	let mut ext = TestExternalities::new(t);
+	ext.execute_with(|| System::set_block_number(1));
+	ext
+}
+
+pub struct CurrencyIdConvert;
+impl Convert<CurrencyId, Option<MultiLocation>> for CurrencyIdConvert {
+	fn convert(id: CurrencyId) -> Option<MultiLocation> {
```

Comment:
Indents look too big

Reply:
Decided to fix it in separate PR

Reply:
created task for that https://github.com/sora-xor/sora2-parachain/issues/74